File: java/009_need-for-speed/NeedForSpeed.java
class NeedForSpeed {
    protected final int speed;
    protected int battery;
    protected int distance;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.battery = (int)100 / batteryDrain;
    }

    public boolean batteryDrained() {
        return battery == 0;
    }

    public int distanceDriven() {
        return distance;
    }

    public void drive() {
        if (!batteryDrained()) {
            distance += speed;
            battery--;
        }     
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50,4);
    }
}

class RaceTrack {
    private final int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return car.battery * car.speed >= distance;
    }
}

File: csharp/009_need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private readonly int _speed;
    private readonly int _batteryDrain;
    private int _currentDistance;
    private int _currentBattery;
    
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this._speed = speed;
        this._batteryDrain = batteryDrain;
        this._currentDistance = 0;
        this._currentBattery = 100;
    }
    
    public bool BatteryDrained()
        => _currentBattery < _batteryDrain;

    public int DistanceDriven()
        => _currentDistance;

    public void Drive()
    {
        if(!BatteryDrained()) {
            _currentDistance += _speed;
            _currentBattery  -= _batteryDrain;
        }
    }

    public static RemoteControlCar Nitro()
        => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private readonly int _distance;
        
    public RaceTrack(int distance)
    {
        this._distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while(!car.BatteryDrained())
        {
            car.Drive();
            if(car.DistanceDriven() >= _distance)
                return true;
        }
        return false;
    }
}